File: Android/Java/app/src/main/java/com/microsoft/sampleandroid/AnchorPlacementFragment.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
package com.microsoft.sampleandroid;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.RadioGroup;
import android.widget.TextView;

import com.google.ar.core.Anchor;
import com.google.ar.core.HitResult;
import com.google.ar.core.Plane;
import com.google.ar.sceneform.ux.ArFragment;

public class AnchorPlacementFragment extends Fragment {
    private ArFragment arFragment;
    private AnchorVisual visual;
    private AnchorPlacementListener listener;

    private TextView hintText;
    private Button confirmPlacementButton;
    private RadioGroup shapeSelection;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.coarse_reloc_anchor_placement, container, false);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        FragmentActivity activity = (FragmentActivity)context;
        arFragment = (ArFragment)activity.getSupportFragmentManager().findFragmentById(R.id.ar_fragment);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        confirmPlacementButton = view.findViewById(R.id.confirm_placement);
        hintText = view.findViewById(R.id.hint_text);
        shapeSelection = view.findViewById(R.id.shape_selection);
    }

    @Override
    public void onStart() {
        super.onStart();

        confirmPlacementButton.setEnabled(false);
        arFragment.setOnTapArPlaneListener(this::onTapArPlaneListener);
        confirmPlacementButton.setOnClickListener(this::onConfirmPlacementClicked);
        shapeSelection.setOnCheckedChangeListener(this::onShapeSelected);
    }

    public void setListener(AnchorPlacementListener listener) {
        this.listener = listener;
    }

    @Override
    public void onStop() {
        arFragment.setOnTapArPlaneListener(null);

        if (visual != null) {
            visual.destroy();
            visual = null;
        }

        super.onStop();
    }

    private void onTapArPlaneListener(HitResult hitResult, Plane plane, MotionEvent motionEvent) {
        if (visual != null) {
            visual.destroy();
            visual = null;
        }

        Anchor localAnchor = hitResult.createAnchor();
        visual = new AnchorVisual(arFragment, localAnchor);
        visual.setMovable(true);
        visual.setShape(getSelectedShape());
        visual.setColor(arFragment.getContext(), android.graphics.Color.YELLOW);
        visual.render(arFragment);

        hintText.setText(R.string.hint_adjust_anchor);

        confirmPlacementButton.setEnabled(true);
    }

    private void onShapeSelected(RadioGroup radioGroup, int selectedId) {
        if (visual == null) {
            return;
        }

        visual.setShape(getSelectedShape());
    }

    private AnchorVisual.Shape getSelectedShape() {
        int selectedId = shapeSelection.getCheckedRadioButtonId();
        if (selectedId == R.id.sphere_shape) {
            return AnchorVisual.Shape.Sphere;
        } else if (selectedId == R.id.cylinder_shape) {
            return AnchorVisual.Shape.Cylinder;
        } else if (selectedId == R.id.cube_shape) {
            return AnchorVisual.Shape.Cube;
        }

        throw new IllegalStateException("Invalid selected shape");
    }

    private void onConfirmPlacementClicked(View view) {
        if (visual != null) {
            AnchorVisual placedAnchor = visual;
            visual = null;
            placedAnchor.setMovable(false);
            if (listener != null) {
                listener.onAnchorPlaced(placedAnchor);
            }
        }
    }
}

File: Xamarin/SampleXamarin.Android/AnchorPlacementFragment.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

using Android.Content;
using Android.OS;
using Android.Support.V4.App;
using Android.Views;
using Android.Widget;
using Google.AR.Core;
using Google.AR.Sceneform.UX;
using System;
using static Android.Views.View;
using static Android.Widget.RadioGroup;
using static Google.AR.Sceneform.UX.BaseArFragment;
using Color = Android.Graphics.Color;

namespace SampleXamarin
{
    internal class AnchorPlacementFragment
        : Fragment, IOnTapArPlaneListener, IOnClickListener, IOnCheckedChangeListener
    {
        private ArFragment arFragment;
        private AnchorVisual visual;

        private TextView hintText;
        private Button confirmPlacementButton;
        private RadioGroup shapeSelection;

        private AnchorVisual.NamedShape SelectedShape
        {
            get
            {
                int selectedId = shapeSelection.CheckedRadioButtonId;
                if (selectedId == Resource.Id.sphere_shape)
                {
                    return AnchorVisual.NamedShape.Sphere;
                }
                else if (selectedId == Resource.Id.cylinder_shape)
                {
                    return AnchorVisual.NamedShape.Cylinder;
                }
                else if (selectedId == Resource.Id.cube_shape)
                {
                    return AnchorVisual.NamedShape.Cube;
                }

                throw new InvalidOperationException("Invalid selected shape");
            }
        }

        public delegate void AnchorPlacementListener(AnchorVisual visual);

        public AnchorPlacementListener OnAnchorPlaced { private get; set; }

        public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
        {
            return inflater.Inflate(Resource.Layout.coarse_reloc_anchor_placement, container, false);
        }

        public override void OnAttach(Context context)
        {
            base.OnAttach(context);

            FragmentActivity activity = (FragmentActivity)context;
            arFragment = (ArFragment)activity.SupportFragmentManager.FindFragmentById(Resource.Id.ar_fragment);
        }

        public override void OnViewCreated(View view, Bundle savedInstanceState)
        {
            base.OnViewCreated(view, savedInstanceState);

            hintText = view.FindViewById<TextView>(Resource.Id.hint_text);
            confirmPlacementButton = view.FindViewById<Button>(Resource.Id.confirm_placement);
            shapeSelection = view.FindViewById<RadioGroup>(Resource.Id.shape_selection);
        }

        public override void OnStart()
        {
            base.OnStart();

            confirmPlacementButton.Enabled = false;
            arFragment.SetOnTapArPlaneListener(this);
            confirmPlacementButton.SetOnClickListener(this);
            shapeSelection.SetOnCheckedChangeListener(this);
        }

        public override void OnStop()
        {
            arFragment.SetOnTapArPlaneListener(null);

            if (visual != null)
            {
                visual.Destroy();
                visual = null;
            }

            base.OnStop();
        }

        public void OnTapPlane(HitResult hitResult, Plane plane, MotionEvent motionEvent)
        {
            if (visual != null)
            {
                visual.Destroy();
                visual = null;
            }

            Anchor localAnchor = hitResult.CreateAnchor();
            visual = new AnchorVisual(arFragment, localAnchor);
            visual.IsMovable = true;
            visual.Shape = SelectedShape;
            visual.SetColor(arFragment.Context, Color.Yellow);
            visual.AddToScene(arFragment);

            hintText.SetText(Resource.String.hint_adjust_anchor);

            confirmPlacementButton.Enabled = true;
        }

        void IOnCheckedChangeListener.OnCheckedChanged(RadioGroup radioGroup, int selectedId)
        {
            if (visual == null)
            {
                return;
            }

            visual.Shape = SelectedShape;
        }

        public void OnClick(View view)
        {
            if (visual != null)
            {
                AnchorVisual placedAnchor = visual;
                visual = null;
                placedAnchor.IsMovable = false;
                OnAnchorPlaced?.Invoke(placedAnchor);
            }
        }
    }
}
